### PR TITLE
`ScheduleForm`, `TemplateForm`, `ReportForm`, and `TimeFrameForm` Refactoring

### LINKED ISSUE #164: Enhance usage of multiple buttons in a form
## Describe the bug

How we currently use the ipl form here is way too complicated and especially having more than one submit button in a form. Most of them are quite old from 2019 and at the time our ipl-html and ipl-web had been quite new :). But now we can remove all the tricks of deleting entries while the form is assembled things... and can be implement in a more elegant way instead.

### Implementation Tips

If a form has more than one Submit button, you can simply override the `hasBeenSubmitted()` method and check by yourself whether the concrete form is Submitted or not. For example, if you have a submit button and a remove button, the override method could look like this.

```php
public function hasBeenSubmitted()
{
        return $this->hasBeenSent() && ($this->getPopulatedValue('submit') || $this->getPopulatedValue('remove'));
}
```

### Affected forms

- `ReportForm`, `ScheduleForm`, `TemplateForm` and `TimeFrameForm`
- And all the controllers which uses these forms

### PR DESCRIPTION
**Pull request by TAINCER**
This removes the Multiple Submit Buttons Hacks from `ScheduleForm`, `ReportForm`, `TemplateForm`, and `TimeFrameForm`. it also adds PHPDocs and Typo annotation where possible.

fixes #164

### CODE REVIEW

File: library/Reporting/Web/Forms/TemplateForm.php [hunk 1]
```diff
@@ -16,30 +16,47 @@ class TemplateForm extends CompatForm
...
             }));
         }
 
-        return $this;
+        return $form;
+    }
+
+    /**
+     * @inheritDoc
+     */
```
Comment (yhabteab) [inline library/Reporting/Web/Forms/TemplateForm.php:51]: We use `inheritDoc` only when adding custom ones in addition to the parent docs.

File: library/Reporting/Web/Forms/TemplateForm.php [hunk 2]
```diff
@@ -16,30 +16,47 @@ class TemplateForm extends CompatForm
...
+    public function hasBeenSubmitted(): bool
+    {
+        return $this->hasBeenSent() && ($this->getPopulatedValue('submit') || $this->getPopulatedValue('remove'));
     }
 
+    /**
+     * @inheritDoc
+     */
```
Comment (yhabteab) [inline library/Reporting/Web/Forms/TemplateForm.php:59]: same here!

File: library/Reporting/Web/Forms/TemplateForm.php [hunk 3]
```diff
@@ -131,24 +148,13 @@ protected function assemble()
...
-            }
         }
     }
 
     public function onSuccess()
     {
-        if ($this->callOnSuccess === false) {
-            return;
+        if ($this->getPopulatedValue('remove')) {
+            $this->getDb()->delete('template', ['id = ?' => $this->template->id]);
```
Comment (yhabteab) [inline library/Reporting/Web/Forms/TemplateForm.php:157]: Forget to return here?

**Review by yhabteab — CHANGES_REQUESTED** [review]
That's still being used though!

https://github.com/Icinga/icingaweb2-module-reporting/blob/17600800bbb00f5160df50476d8fceb9b933ca51/library/Reporting/Web/Controller.php#L12-L19

File: library/Reporting/Web/Forms/ScheduleForm.php [hunk 4]
```diff
@@ -25,36 +26,56 @@ class ScheduleForm extends CompatForm
...
 
-    public function setReport(Report $report)
+    /**
+     * Creates a new `ScheduleForm` instance with a given Report
+     *
+     * @param Report $report
+     *
+     * @return ScheduleForm
+     */
+    public static function fromReport(Report $report): ScheduleForm
```
[the remark this answers is not in the document]
Comment (lippserd) [reply library/Reporting/Web/Forms/ScheduleForm.php:39]: Fix that for the other forms as well please.

File: library/Reporting/Web/Forms/ScheduleForm.php [hunk 5]
```diff
@@ -25,36 +26,56 @@ class ScheduleForm extends CompatForm
     /** @var Report */
     protected $report;
 
+    /** @var int */
     protected $id;
 
-    public function setReport(Report $report)
+    /**
+     * Creates a new `ScheduleForm` instance with a given Report
```
[the remark this answers is not in the document]
Comment (lippserd) [reply library/Reporting/Web/Forms/ScheduleForm.php:33]: Fix that for the other forms as well please.

**Review by lippserd — CHANGES_REQUESTED** [review]

**Comment by lippserd** [comment]
@yhabteab Do you want to have a final look here?

**Review by yhabteab — APPROVED** [review]
LFTM as well! :)

**Pull request merged**